File: openjaus/trunk/Core/libjausC/src/message/experimental/events/confirmEventMessage.c
// File Name: confirmEventMessage.c
//
// Written By: Danny Kent (jaus AT dannykent DOT com), Tom Galluzzo (galluzzo AT gmail DOT com)
//
// Version: 3.2
//
// Date: 08/04/06
//
// Description: This file defines the functionality of a ConfirmEventMessage



#include <stdlib.h>
#include <string.h>
#include <cimar.h>
#include "cimar/jaus.h"

static const int commandCode = JAUS_CONFIRM_EVENT;
static const int maxDataSizeBytes = 7;

static JausBoolean headerFromBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes);
static JausBoolean headerToBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes);

static JausBoolean dataFromBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes);
static int dataToBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes);
static void dataInitialize(ConfirmEventMessage message);
static void dataDestroy(ConfirmEventMessage message);

// ************************************************************************************************************** //
//                                    USER CONFIGURED FUNCTIONS
// ************************************************************************************************************** //

// Initializes the message-specific fields
static void dataInitialize(ConfirmEventMessage message)
{
	message->expFlag = JAUS_EXPERIMENTAL_MESSAGE;	
	// Set initial values of message fields
	message->presenceVector = newJausBytePresenceVector();	// 1: Presence Vector
	message->messageCode = newJausUnsignedShort(0);			// 2: Command Code of message sent in case of event occurance
	message->eventId = newJausByte(0);						// 3: Event Id
	message->confirmedUpdateRate = newJausDouble(0);		// 4: For Periodic Events, Scaled UnsignedShort (0, 1092)
	message->responseCode = newJausByte(0);					// 5: Enumeration of Response Types (see above)
}

// Destructs the message-specific fields
static void dataDestroy(ConfirmEventMessage message)
{
	// Free message fields
}

// Return boolean of success
static JausBoolean dataFromBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	int index = 0;
	JausUnsignedShort tempUShort;
	
	if(bufferSizeBytes == message->dataSize)
	{
		// Unpack Message Fields from Buffer
		// Presence Vector
		if(!jausBytePresenceVectorFromBuffer(&message->presenceVector, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_PRESENCE_VECTOR_SIZE_BYTES;

		// Message Code
		if(!jausUnsignedShortFromBuffer(&message->messageCode, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_UNSIGNED_SHORT_SIZE_BYTES;		

		// Event Id
		if(!jausByteFromBuffer(&message->eventId, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_SIZE_BYTES;

		if(jausBytePresenceVectorIsBitSet(message->presenceVector, CONFIRM_EVENT_PV_PERIODIC_RATE_BIT))
		{		
			// Periodic Rate 
			if(!jausUnsignedShortFromBuffer(&tempUShort, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
			index += JAUS_UNSIGNED_SHORT_SIZE_BYTES;
			
			// Scaled Int (0, 1092);
			message->confirmedUpdateRate = jausUnsignedShortToDouble(tempUShort, 0, 1092);
		}
		
		// Response Code
		if(!jausByteFromBuffer(&message->responseCode, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_SIZE_BYTES;

		return JAUS_TRUE;
	}
	else
	{
		return JAUS_FALSE;
	}
}

// Returns number of bytes put into the buffer
static int dataToBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	int index = 0;
	JausUnsignedShort tempUShort;

	if(bufferSizeBytes >= message->dataSize)
	{
		// Pack Message Fields to Buffer
		// Presence Vector
		if(!jausBytePresenceVectorToBuffer(message->presenceVector, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_PRESENCE_VECTOR_SIZE_BYTES;

		// Message Code
		if(!jausUnsignedShortToBuffer(message->messageCode, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_UNSIGNED_SHORT_SIZE_BYTES;		

		// Event Id
		if(!jausByteToBuffer(message->eventId, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_SIZE_BYTES;

		if(jausBytePresenceVectorIsBitSet(message->presenceVector, CONFIRM_EVENT_PV_PERIODIC_RATE_BIT))
		{		
			// Scaled Int (0, 1092);
			tempUShort = jausUnsignedShortToDouble(message->confirmedUpdateRate, 0, 1092);

			// Periodic Rate 
			if(!jausUnsignedShortToBuffer(tempUShort, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
			index += JAUS_UNSIGNED_SHORT_SIZE_BYTES;
		}

		// Response Code
		if(!jausByteToBuffer(message->responseCode, buffer+index, bufferSizeBytes-index)) return JAUS_FALSE;
		index += JAUS_BYTE_SIZE_BYTES;		
	}

	return index;
}

// ************************************************************************************************************** //
//                                    NON-USER CONFIGURED FUNCTIONS
// ************************************************************************************************************** //

ConfirmEventMessage confirmEventMessageCreate(void)
{
	ConfirmEventMessage message;

	message = (ConfirmEventMessage)malloc( sizeof(ConfirmEventMessageStruct) );
	if(message == NULL)
	{
		return NULL;
	}
	
	// Initialize Values
	message->priority = JAUS_DEFAULT_PRIORITY;
	message->ackNak = JAUS_ACK_NAK_NOT_REQUIRED;
	message->scFlag = JAUS_NOT_SERVICE_CONNECTION_MESSAGE;
	message->expFlag = JAUS_NOT_EXPERIMENTAL_MESSAGE;
	message->version = JAUS_VERSION_3_2;
	message->reserved = 0;
	message->commandCode = commandCode;
	message->destination = jausAddressCreate();
	message->source = jausAddressCreate();
	message->dataFlag = JAUS_SINGLE_DATA_PACKET;
	message->dataSize = maxDataSizeBytes;
	message->sequenceNumber = 0;
	
	dataInitialize(message);
	
	return message;	
}

void confirmEventMessageDestroy(ConfirmEventMessage message)
{
	dataDestroy(message);
	jausAddressDestroy(message->source);
	jausAddressDestroy(message->destination);
	free(message);
}

JausBoolean confirmEventMessageFromBuffer(ConfirmEventMessage message, unsigned char* buffer, unsigned int bufferSizeBytes)
{
	int index = 0;
	
	if(!strncmp((char *)buffer, JAUS_UDP_HEADER, JAUS_UDP_HEADER_SIZE_BYTES)) // equals 1 if same
	{
		index += JAUS_UDP_HEADER_SIZE_BYTES;
	}

	if(headerFromBuffer(message, buffer+index, bufferSizeBytes-index))
	{
		index += JAUS_HEADER_SIZE_BYTES;
		if(dataFromBuffer(message, buffer+index, bufferSizeBytes-index))
		{
			return JAUS_TRUE;
		}
		else
		{
			return JAUS_FALSE;
		}
	}
	else
	{
		return JAUS_FALSE;
	}
}

JausBoolean confirmEventMessageToBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	if(bufferSizeBytes < confirmEventMessageSize(message))
	{
		return JAUS_FALSE; //improper size	
	}
	else
	{	
		message->dataSize = dataToBuffer(message, buffer+JAUS_HEADER_SIZE_BYTES, bufferSizeBytes - JAUS_HEADER_SIZE_BYTES);
		if(headerToBuffer(message, buffer, bufferSizeBytes))
		{
			return JAUS_TRUE;
		}
		else
		{
			return JAUS_FALSE; // headerToConfirmEventBuffer failed
		}
	}
}

JausBoolean confirmEventMessageToUdpBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	if(bufferSizeBytes < confirmEventMessageUdpSize(message))
	{
		return JAUS_FALSE; // improper size
	}
	else
	{
		strncpy( (char *) buffer, JAUS_UDP_HEADER, JAUS_UDP_HEADER_SIZE_BYTES); //copies the UDP header into the buffer
		return confirmEventMessageToBuffer(message, buffer+JAUS_UDP_HEADER_SIZE_BYTES, bufferSizeBytes - JAUS_UDP_HEADER_SIZE_BYTES);
	}
}

ConfirmEventMessage confirmEventMessageFromJausMessage(JausMessage jausMessage)
{
	ConfirmEventMessage message;
	
	if(jausMessage->commandCode != commandCode)
	{
		return NULL; // Wrong message type
	}
	else
	{
		message = (ConfirmEventMessage)malloc( sizeof(ConfirmEventMessageStruct) );
		if(message == NULL)
		{
			return NULL;
		}
		
		message->properties = jausMessage->properties;
		message->commandCode = jausMessage->commandCode;
		message->destination = jausAddressCreate();
		message->destination->id = jausMessage->destination->id;
		message->source = jausAddressCreate();
		message->source->id = jausMessage->source->id;
		message->dataControl = jausMessage->dataControl;
		message->sequenceNumber = jausMessage->sequenceNumber;
		
		// Unpack jausMessage->data
		if(dataFromBuffer(message, jausMessage->data, jausMessage->dataSize))
		{
			return message;
		}
		else
		{
			return NULL;
		}
	}
}

JausMessage confirmEventMessageToJausMessage(ConfirmEventMessage message)
{
	JausMessage jausMessage;
	
	jausMessage = (JausMessage)malloc( sizeof(JausMessageStruct) );
	if(jausMessage == NULL)
	{
		return NULL;
	}	
	
	jausMessage->properties = message->properties;
	jausMessage->commandCode = message->commandCode;
	jausMessage->destination = jausAddressCreate();
	jausMessage->destination->id = message->destination->id;
	jausMessage->source = jausAddressCreate();
	jausMessage->source->id = message->source->id;
	jausMessage->dataControl = message->dataControl;
	jausMessage->sequenceNumber = message->sequenceNumber;
	
	jausMessage->data = (unsigned char *)malloc(message->dataSize);
	jausMessage->dataSize = dataToBuffer(message, jausMessage->data, message->dataSize);
	
	return jausMessage;
}

unsigned int confirmEventMessageUdpSize(ConfirmEventMessage message)
{
	return (unsigned int)(message->dataSize + JAUS_HEADER_SIZE_BYTES + JAUS_UDP_HEADER_SIZE_BYTES);
}

unsigned int confirmEventMessageSize(ConfirmEventMessage message)
{
	return (unsigned int)(message->dataSize + JAUS_HEADER_SIZE_BYTES);
}

//********************* PRIVATE HEADER FUNCTIONS **********************//

static JausBoolean headerFromBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	if(bufferSizeBytes < JAUS_HEADER_SIZE_BYTES)
	{
		return JAUS_FALSE;
	}
	else
	{
		// unpack header
		message->properties = buffer[0] + (buffer[1] << 8);
		message->commandCode = buffer[2] + (buffer[3] << 8);
	
		message->destination->instance = buffer[4];
		message->destination->component = buffer[5];
		message->destination->node = buffer[6];
		message->destination->subsystem = buffer[7];
	
		message->source->instance = buffer[8];
		message->source->component = buffer[9];
		message->source->node = buffer[10];
		message->source->subsystem = buffer[11];
		
		message->dataControl = buffer[12] + (buffer[13] << 8);
		message->sequenceNumber = buffer[14] + (buffer[15] << 8);
		
		return JAUS_TRUE;
	}
}

static JausBoolean headerToBuffer(ConfirmEventMessage message, unsigned char *buffer, unsigned int bufferSizeBytes)
{
	if(bufferSizeBytes < JAUS_HEADER_SIZE_BYTES)
	{
		return JAUS_FALSE;
	}
	else
	{	
		buffer[0] = (unsigned char)(message->properties & 0xFF);
		buffer[1] = (unsigned char)((message->properties & 0xFF00) >> 8);

		buffer[2] = (unsigned char)(message->commandCode & 0xFF);
		buffer[3] = (unsigned char)((message->commandCode & 0xFF00) >> 8);

		buffer[4] = (unsigned char)(message->destination->instance & 0xFF);
		buffer[5] = (unsigned char)(message->destination->component & 0xFF);
		buffer[6] = (unsigned char)(message->destination->node & 0xFF);
		buffer[7] = (unsigned char)(message->destination->subsystem & 0xFF);

		buffer[8] = (unsigned char)(message->source->instance & 0xFF);
		buffer[9] = (unsigned char)(message->source->component & 0xFF);
		buffer[10] = (unsigned char)(message->source->node & 0xFF);
		buffer[11] = (unsigned char)(message->source->subsystem & 0xFF);
		
		buffer[12] = (unsigned char)(message->dataControl & 0xFF);
		buffer[13] = (unsigned char)((message->dataControl & 0xFF00) >> 8);

		buffer[14] = (unsigned char)(message->sequenceNumber & 0xFF);
		buffer[15] = (unsigned char)((message->sequenceNumber & 0xFF00) >> 8);
		
		return JAUS_TRUE;
	}
}

File: trunk/Core/libjausJava/src/jaus/message/experimental/ConfirmEventMessage.java
/**
 * @author Tom Galluzzo (galluzzo@gmail.com)
 * @version 0.1
 * @date: 02/14/05
 */

package jaus.message.experimental;
import jaus.message.*;
import jaus.type.*;

public class ConfirmEventMessage extends JausMessage
{
	private static final int COMMAND_CODE = CONFIRM_EVENT;
	private static final int MAX_DATA_SIZE_BYTES = 7;

	public static final int CONFIRMED_UPDATE_RATE_BIT = 0;
	
	//Message Fields
	JausBytePresenceVector presenceVector;	// 1: Presence Vector
	JausUnsignedShort messageCode;			// 2: Command Code of message sent in case of event occurance
	JausByte eventId;						// 3: Event Id
	JausUnsignedShort confirmedUpdateRate;			// 4: For Periodic Events, Scaled UnsignedShort (0, 1092)
	JausByte responseCode;					// 5: Enumeration of Response Types (see above)

	public ConfirmEventMessage()
	{
		super();
		setUserDefinedFlag(EXPERIMENTAL_MESSAGE);

		data = new byte[MAX_DATA_SIZE_BYTES];

		presenceVector = new JausBytePresenceVector();	// 1: Presence Vector
		messageCode = new JausUnsignedShort(0);			// 2: Command Code of the resulting query
		eventId = new JausByte(0);
		confirmedUpdateRate = new JausUnsignedShort(0);	
		responseCode = new JausByte(0);
	}
	
	public ConfirmEventMessage(byte[] buffer)
	{
	    setFromJausBuffer(buffer, 0);
	}
	
	public ConfirmEventMessage(byte[] buffer, int index)
	{
	    setFromJausBuffer(buffer, index);
	}

	public ConfirmEventMessage(JausMessage jausMessage)
	{
		super(jausMessage);
		int dataIndex = 0;

		presenceVector = new JausBytePresenceVector(data, dataIndex) ;
		dataIndex += presenceVector.size();
		
		messageCode = new JausUnsignedShort(data, dataIndex);
		dataIndex += messageCode.size();
		
		eventId = new JausByte(data, dataIndex);
		dataIndex += eventId.size();
		
		if(presenceVector.isBitSet(CONFIRMED_UPDATE_RATE_BIT))
		{
			confirmedUpdateRate = new JausUnsignedShort(data, dataIndex);
			dataIndex += confirmedUpdateRate.size();
		}

		responseCode = new JausByte(data, dataIndex);
		dataIndex += responseCode.size();
	}

	public void setMessageCode(int messageCode)
	{
		this.messageCode = new JausUnsignedShort(messageCode);
	}
	
	public void setResponseCode(int code)
	{
		this.responseCode = new JausByte(code);
	}
	
	public void setCommandCode(int commandCode)
	{
	    this.commandCode = COMMAND_CODE;
	}	

	public boolean setPresenceVectorBit(int bit)
	{
		return presenceVector.setBit(bit);
	}
	
	public boolean clearPresenceVectorBit(int bit)
	{
		return presenceVector.clearBit(bit);
	}
	
	// Takes a data UDP or regular JAUS buffer and unpacks it into the header data structure and the data attributes
	public boolean setFromJausBuffer(byte[] buffer, int index)
	{
		if(super.setFromJausBuffer(buffer, index))
		{
			int dataIndex = 0;

			presenceVector = new JausBytePresenceVector(data, dataIndex) ;
			dataIndex += presenceVector.size();
			
			messageCode = new JausUnsignedShort(data, dataIndex);
			dataIndex += messageCode.size();
			
			eventId = new JausByte(data, dataIndex);
			dataIndex += eventId.size();
			
			if(presenceVector.isBitSet(CONFIRMED_UPDATE_RATE_BIT))
			{
				confirmedUpdateRate = new JausUnsignedShort(data, dataIndex);
				dataIndex += confirmedUpdateRate.size();
			}

			responseCode = new JausByte(data, dataIndex);
			dataIndex += responseCode.size();

			return true;
		}
		else
		{
		    return false;
		}
	}
	
	// Takes the header and data byte array and packs them into a data buffer
	// This method needs to be overridden for all subclasses of JausMessage to reflect the correct pack routine
	public boolean toJausBuffer(byte[] buffer, int index)
	{
	    if(super.toJausBuffer(buffer, index))
	    {
			index += HEADER_SIZE_BYTES;

			if(presenceVector.getPresenceVector().toJausBuffer(buffer, index))
			{
				index += presenceVector.size();
			}
			else
			{
				return false;
			}

			if(messageCode.toJausBuffer(buffer, index))
			{
				index += messageCode.size();
			}
			else
			{
				return false;
			}

			if(eventId.toJausBuffer(buffer, index))
			{
				index += eventId.size();
			}
			else
			{
				return false;
			}

			if(presenceVector.isBitSet(CONFIRMED_UPDATE_RATE_BIT))
			{
				if(confirmedUpdateRate.toJausBuffer(buffer, index))
				{
					index += confirmedUpdateRate.size();
				}
				else
				{
					return false;
				}
			}			

			if(responseCode.toJausBuffer(buffer, index))
			{
				index += responseCode.size();
			}
			else
			{
				return false;
			}
		
			return true;
	    }
	    else
	    {
	        return false;
	    }
	}

	// Overloaded method to accept a buffer and pack this message into its UDP form
	public boolean toJausUdpBuffer(byte[] buffer, int index)
	{
	    if(super.toJausUdpBuffer(buffer, index))
	    {
			index += HEADER_SIZE_BYTES + UDP_HEADER_SIZE_BYTES;

			if(presenceVector.getPresenceVector().toJausBuffer(buffer, index))
			{
				index += presenceVector.size();
			}
			else
			{
				return false;
			}

			if(messageCode.toJausBuffer(buffer, index))
			{
				index += messageCode.size();
			}
			else
			{
				return false;
			}

			if(eventId.toJausBuffer(buffer, index))
			{
				index += eventId.size();
			}
			else
			{
				return false;
			}

			if(presenceVector.isBitSet(CONFIRMED_UPDATE_RATE_BIT))
			{
				if(confirmedUpdateRate.toJausBuffer(buffer, index))
				{
					index += confirmedUpdateRate.size();
				}
				else
				{
					return false;
				}
			}			

			if(responseCode.toJausBuffer(buffer, index))
			{
				index += responseCode.size();
			}
			else
			{
				return false;
			}
			
			return true;
	    }
	    else
	    {
	        return false;
	    }
	}

}